Add shellcheck to CI - Closes #1685

khanku opened the pull request:
Run shellcheck in docker on Travis against all `*.sh` files under `dev/scripts/` and all files with a sh/bash/dash/ksh shebang under `misc/userscripts/`.
Closes #1685 

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3355)
<!-- Reviewable:end -->

jgkamat commented on scripts/dev/download_release.sh:
```diff
@@ -11,7 +11,7 @@ if [[ $# != 1 ]]; then
     exit 1
 fi
 
-cd "$tmpdir"
```
Instead of adding all these `|| exit`'s, can we `set -e` at the top of the file to exit on any failing command?

khanku replied:
Sure, that makes sense. I'll do that.

khanku commented on misc/userscripts/format_json:
```diff
@@ -22,25 +22,24 @@
 
 # calculate the filesize of the json document
+# parsing the output of ls should be fine in this case since we set the block size
+# shellcheck disable=SC2012
```
Yes, that should be easy enough.

khanku commented:
Regarding koalaman/shellcheck#785: I switched to the `koalaman/shellcheck:latest` docker image and removed a `shellcheck disable` comment: shellcheck now complains as expected (see this [build log](https://travis-ci.org/khanku/qutebrowser/jobs/310774986))

jgkamat commented on misc/userscripts/format_json:
```diff
@@ -19,29 +20,23 @@
+# try GNU stat first and then OSX stat if the former fails
+FILE_SIZE=$(
+  stat --printf="%s" "$TEMP_FILE" 2>/dev/null ||
+  stat -f%z "$TEMP_FILE" 2>/dev/null
```
It's probably a good idea to exit after this line just in case no `stat` is found.

khanku replied:
I do not think that's necessary since `set -e` is in effect.

jgkamat replied:
ah, my bad, ignore me :)

jgkamat commented on misc/userscripts/format_json:
```diff
@@ -19,29 +20,23 @@
+  stat -f%z "$TEMP_FILE" 2>/dev/null
+)
+if [ "$FILE_SIZE" -lt "$MAX_SIZE_PRETTIFY" ]; then
+    pygmentize -l json -f html -O full,style="$STYLE" <"$TEMP_FILE" >"${TEMP_FILE}_"
```
pygmentize has an `-o` flag for out file, maybe specifying the same in and out file will work? Not sure about that though, so if not, no worries.

khanku replied:
I'll give that flag a try, thanks for the hint.

khanku replied:
Correct, which is why I reverted that change.

khanku commented:
Actually I'll revert this last commit as it would be a potential race condition... as shellcheck correctly points out ^^

The-Compiler commented on misc/userscripts/rss:
```diff
@@ -85,7 +85,7 @@ if [ ! -d "$config_dir/read_urls" ]; then
 fi
 
-cd "$config_dir"
+cd "$config_dir" || exit
```
`exit 1 ` probably?

The-Compiler commented on scripts/dev/ci/travis_run.sh:
```diff
@@ -14,6 +14,16 @@ elif [[ $TESTENV == eslint ]]; then
     # travis env
     cd qutebrowser/javascript || exit 1
     eslint --color --report-unused-disable-directives .
+elif [[ $TESTENV == shellcheck ]]; then
```
This all seems like overcomplicated and like it'd break with filenames with spaces in them - why not simply have newline-separated filenames (i.e. without `-print0` and `xargs` involved) and use `IFS="\n"` instead?

khanku replied:
You're right, it would filenames containing spaces would probably break this contraption. I will rework it.

The-Compiler commented:
@gilbertw1 Can you review the changes to `format_json` please?

@t-wissmann Can you review the changes to `password_fill` and `view_in_mpv`? Some of those things might be from #1697 so maybe you already saw some of the changes.

The-Compiler commented on scripts/dev/ci/travis_run.sh:
```diff
@@ -14,6 +14,16 @@ elif [[ $TESTENV == eslint ]]; then
+    IFS=" " read -r -a scripts <<< "$dev_scripts $userscripts"
+    docker run \
+           -v "$PWD:/outside" \
+	   -w /outside \
```
Wrong indentation?

The-Compiler commented:
I've talked to @t-wissmann in IRC and he said his scripts look good from a quick look, and I can't spot anything bad either - so let's merge this! @gilbertw1 if you still have comments, of course feel free to add them.

Thanks @khanku for the PR, @neeasade for the original implementation, and @jgkamat for all the reviewing! 👍
